doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,94 @@
+- Feature name: Envelope upsert with order by timestamp (limited scope)
+- Associated: https://github.com/MaterializeInc/materialize/issues/16512
+
+# Summary
+
+- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
+- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.
+
+# Scope
+
+The scope will be limited to only allow `ORDER BY TIMESTAMP` (ascending order).
+
+- For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.
+- Limiting it to only `TIMESTAMP` because
+    - `OFFSET` is already default. We can make it explicit though if we want.
+    - `PARTITION`, `KEY`, `TOPIC`, would not make a good order by clause.
```

Comment:
A bit more of an explanation about why these fields are not good order by clauses would be helpful!

```suggestion
    - `PARTITION`, `KEY`, `TOPIC`, would not make for meaningful orderings:
        - The `PARTITION` of a record is typically derived from its key, and so is the same for every record with a given key.
        - `KEY` is, by definition, the same for every record with a given key.
        - `TOPIC` is the same for every record in the topic.
```